allennlp_rc/models/transformer_qa.py

```diff
@@ -0,0 +1,233 @@
+                context_tokens_for_question = metadata_entry["context_tokens"]
+                context_tokens.append(context_tokens_for_question)
+
+                best_span -= 1 + len(metadata_entry["question_tokens"]) + 2
+                assert np.all(best_span >= 0)
+
+                predicted_start, predicted_end = tuple(best_span)
+                assert predicted_start >= 0
+                assert predicted_end < len(context_tokens_for_question)
+
+                while (
+                    predicted_start >= 0
+                    and context_tokens_for_question[predicted_start].idx is None
+                ):
+                    predicted_start -= 1
+                if predicted_start < 0:
+                    logger.warning(
+                        f"Could not map the token '{context_tokens_for_question[best_span[0]].text}' at index "
+                        f"'{best_span[0]}' to an offset in the original text. Did you configure your tokenizer "
+                        'with "calculate_character_offsets": true?'
+                    )
+                    character_start = 0
+                else:
+                    character_start = context_tokens_for_question[predicted_start].idx
+
+                while (
+                    predicted_end < len(context_tokens_for_question)
+                    and context_tokens_for_question[predicted_end].idx is None
+                ):
+                    predicted_end += 1
+                if predicted_end >= len(context_tokens_for_question):
+                    logger.warning(
+                        f"Could not map the token '{context_tokens_for_question[best_span[1]].text}' at index "
+                        f"'{best_span[1]}' to an offset in the original text. Did you configure your tokenizer "
+                        'with "calculate_character_offsets": true?'
+                    )
+                    character_end = len(metadata_entry["context"])
+                else:
+                    end_token = context_tokens_for_question[predicted_end]
+                    character_end = end_token.idx + len(sanitize_wordpiece(end_token.text))
+
+                best_span_string = metadata_entry["context"][character_start:character_end]
+                output_dict["best_span_str"].append(best_span_string)
+
+                answers = metadata_entry.get("answers")
+                self._metrics(best_span_string, answers)
+            output_dict["context_tokens"] = context_tokens
+        return output_dict
+
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        exact_match, f1_score = self._metrics.get_metric(reset)
+        return {
+            "start_acc": self._span_start_accuracy.get_metric(reset),
+            "end_acc": self._span_end_accuracy.get_metric(reset),
+            "span_acc": self._span_accuracy.get_metric(reset),
+            "em": exact_match,
+            "f1": f1_score,
```

Comment:
If there aren't actual EM and F1, because of the duplicate instance issue mentioned earlier (and looking above, I don't think they are), then I would definitely recommend having this be called something like "approx_f1" and "approx_em", so a person isn't tempted to compare these numbers to actual F1 scores.

Reply:
It's now called `_per_instance_metrics`, from a previous comment.

Reply:
Sorry, that does't apply here. I fixed the names in the dictionary in db7d4959fc19aad49b2ffad1e40a1b5247339b88.